### PR TITLE
feat: support for recording internal audio in rooted devices

### LINKED ISSUE #281: add support for internal audio (ROOT)
### Describe your suggested feature

in screen recorder there is no internal audio support, it could be done without root in older Android versions but latest Android versions need root or maybe some form of magisk/kernelsu module.

### Other details

I have seen this [app](https://gitlab.com/vijai/screenrecorder) do it with its [magisk module](https://gitlab.com/vijai/screencam-magisk-module) but in my phone (Samsung Galaxy M31) it doesn't work 

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by SuhasDissa**
closes #281

### Demo recording

https://github.com/you-apps/RecordYou/assets/64766434/3b40bb55-9be7-4683-a1e1-0c9fa9dc0289

<hr>

<details>
  <summary> Guide for Root users</summary>

For this to work the app needs the `android.permission.CAPTURE_AUDIO_OUTPUT` permission. This permission is a privileged permission. so the app needs to be installed as a privileged app.

### How to manually grant the permission

1. Install the app as a normal app
2. use `adb root` or any other method to move the app's apk from `/data/app/<app folder>` to `/system/priv-app/RecordYou/RecordYou.apk`
3. create a new file `priv-app-permission-recordyou.xml` (name doesn't matter) with the following content:
```xml
<permissions>
	<privapp-permissions package="com.bnyro.recorder">
		<permission name="android.permission.CAPTURE_AUDIO_OUTPUT" />
	</privapp-permissions>
</permissions>
```
4. reboot the phone and use the feature
</details>

### CODE REVIEW

**Comment by SuhasDissa** [comment]
Created a magisk module to enable the feature: [RecordYou.zip](https://github.com/you-apps/RecordYou/files/14901944/RecordYou.zip)


> [!WARNING]
> Use at your own risk.

**Comment by SuhasDissa** [comment]
> I think it's not working on Samsung oneUI. but the magisk module is working fine 

Debug builds won't work

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -85,4 +85,6 @@
     <string name="cant_access_selected_folder">Can\'t access selected folder!</string>
     <string name="draw_mode">Draw Mode</string>
     <string name="erase_mode">Erase Mode</string>
+    <string name="internal_audio">Internal Audio (Root)</string>
+    <string name="no_enough_permissions">No enough permissions to start recording</string>
```
Comment (Bnyro) [inline app/src/main/res/values/strings.xml:89]:
```suggestion
    <string name="no_sufficient_permissions">No sufficient permissions to start recording</string>
```

File: app/src/main/java/com/bnyro/recorder/ui/models/RecorderModel.kt [hunk 2]
```diff
@@ -161,12 +179,29 @@ class RecorderModel : ViewModel() {
...
+        if (granted) {
+            return true
+        } else {
+            Toast.makeText(
+                context,
+                context.getString(R.string.no_enough_permissions), Toast.LENGTH_SHORT
+            )
+                .show()
+            return false
+        }
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/ui/models/RecorderModel.kt:205]:
```suggestion
        if (!granted) {
            Toast.makeText(
                context,
                context.getString(R.string.no_enough_permissions), Toast.LENGTH_SHORT
            )
                .show()
        }
        
        return granted
```

**Review by Bnyro — COMMENTED** [review]
I don't have a working rooted device to test this at the moment - but the changes look fine (except for some small nitpicks in the comments).

Though I wonder if we shouldn't show a toast or a note when users want to select that option, that they need root and follow some steps to make it work? Because currently, users would just try it without any setup and think it just doesn't work.

**Comment by Bnyro** [comment]
> Nevertheless, I can create a new repo and put the magisk module there with the instructions, if you're ok with it
>
That's fine with me :+1:

**Comment by Robbilie** [comment]
@SuhasDissa I tried the github actions apk and built a local release and signed it with the test key. In both cases the apk was installed in /system/system_ext/priv-app, adb dump of the package shows the permission is granted (even without the permissions xml, i added it later though for completeness). I tried several variations (audio only and screen recorder, wav and no wav, different encodings, some crash the app though) but all files are without any audio. i am using an android emulator (x86_64, automotive, should probably not make any difference though) and I play back sound using spotify and vlc (an mp3 file).

any clues as to what i am doing wrong?

**Comment by Robbilie** [comment]
> > any clues as to what i am doing wrong?
> 
> Did you set the audio source as internal audio?

Yes 😄

**Pull request merged**